### PR TITLE
FOGL-1921:  Add support to insert data processing plugin into south & north

### PR DESCRIPTION
**Pull request by mshariq-nerd**
This PR need to be tested with https://github.com/foglamp/FogLAMP/pull/1269 (now in develop)

#### STEPS TO FOLLOW

1) Open Service modal 

2) Click on the  Applications **+** icon to open the **Add filter wizard**

screen shot 2018-12-07 at 7 07 57 pm

3) Choose filter plugin, Enter name

screen shot 2018-12-28 at 3 19 48 pm

4) Review (update) configuration 

screen shot 2018-12-28 at 3 20 04 pm

5) Create the filter by clicking Done button. It will redirect control to service modal page, and will **add /update  pipeline in the background.** 

6) Filter linked with the service can be seen on service modal page. (`GET /<service name>/pipeline`)

screen shot 2018-12-28 at 3 22 13 pm

> Config for individual filter is being fetched as `<serviceName>_<filterName>` when we click expand on a filter. 
screen shot 2018-12-28 at 3 18 53 pm

### Completed tasks:
1) Add filter(with out or with changed configuration ) through wizard 
2) Update filter configuration on service modal page
3) Delete filter
4) Drag and drop (reordering) of filters

### CODE REVIEW

File: src/app/components/core/south/south-service-modal/south-service-modal.component.ts [hunk 1]
```diff
@@ -328,6 +341,60 @@ export class SouthServiceModalComponent implements OnInit, OnChanges {
...
+  }
+
+  getFilterPipeline() {
+    this.filterService.getFilterPipeline(this.service['name'])
+      .subscribe((data: any) => {
+        this.filterPipeline = data.result.pipeline;
+      },
+        error => {
+          if (error.status === 404) {
+            this.filterPipeline = '';
```
Comment (praveen-garg) [inline src/app/components/core/south/south-service-modal/south-service-modal.component.ts]: I think this should be an empty array `[]` not empty string `''`

File: src/app/components/core/filter/add-filter-wizard/add-filter-wizard.component.ts [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    this.payload.filter_config = finalConfig;
+  }
+
+  /**
+   * Method to add service
+   * @param payload  to pass in request
+   * @param nxtButton button to go next
+   * @param previousButton button to go previous
+   */
+  public addFilter(payload) {
```
Comment (monika-Dianomic) [inline src/app/components/core/filter/add-filter-wizard/add-filter-wizard.component.ts]: Doc comments are not matching with method.

File: src/app/components/core/configuration-manager/view-config-item/view-config-item.component.ts [hunk 3]
```diff
@@ -31,10 +33,11 @@ export class ViewConfigItemComponent implements OnInit, OnChanges {
   ngOnChanges(changes: SimpleChanges) {
     this.filesToUpload = [];
     this.configItems = [];
-    if (changes.categoryConfigurationData.currentValue !== undefined) {
+    const catConfigData = cloneDeep(changes.categoryConfigurationData.currentValue);
```
Comment (praveen-garg) [inline src/app/components/core/configuration-manager/view-config-item/view-config-item.component.ts]: catConfigData > categoryConfigurationCurrentData

File: src/app/components/core/configuration-manager/view-config-item/view-config-item.component.ts [hunk 4]
```diff
@@ -76,6 +80,7 @@ export class ViewConfigItemComponent implements OnInit, OnChanges {
       return this.configItems.map(conf => {
         if (conf.key === d.key) {
           d['type'] = conf.type;
+          d.value = d.value.toString();
```
Comment (praveen-garg) [inline src/app/components/core/configuration-manager/view-config-item/view-config-item.component.ts]: why type in `[]` and value with `.`

File: src/app/components/core/south/south-service-modal/south-service-modal.component.html [hunk 5]
```diff
@@ -5,70 +5,113 @@
   <div class="modal-background"></div>
   <div class="modal-card">
     <header class="modal-card-head">
-      <p *ngIf="service != undefined" class="modal-card-title is-size-5">{{service.name}} South Service</p>
+      <p *ngIf="service != undefined " class="modal-card-title is-size-5">{{service.name}} South Service</p>
```
Comment (praveen-garg) [inline src/app/components/core/south/south-service-modal/south-service-modal.component.html]: why space added after `undefined`

File: src/app/components/core/south/south-service-modal/south-service-modal.component.html [hunk 6]
```diff
@@ -5,70 +5,113 @@
...
+          </div>
+          <div *ngIf="service != undefined" class="columns">
+            <div class="column">
+              <div class="field is-horizontal">
+                <div class="field-label has-text-left">
+                  <label class="label">Enabled</label>
+                </div>
+                <div class="field-body">
+                  <div class="field">
+                    <input [formControl]="svcCheckbox" class="checkbox" type="checkbox" [checked]="service.status !== 'shutdown' && service.status !== '' ">
```
Comment (praveen-garg) [inline src/app/components/core/south/south-service-modal/south-service-modal.component.html]: NOT properly rebased?

File: src/app/services/filter.service.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+   * GET | plugins/installed?type=filter
+   */
+  public getInstalledFilterPlugins() {
+    return this.http.get(this.GET_INSTALLED_FILTER_PLUGINS).pipe(
+      map(response => response),
+      catchError((error: Response) => observableThrowError(error)));
+  }
+
+  /**
+   *  Create filter
```
Comment (praveen-garg) [inline src/app/services/filter.service.ts]: add 1 more new line?

File: src/app/services/filter.service.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+      map(response => response),
+      catchError((error: Response) => observableThrowError(error)));
+  }
+
+
+  /**
+   * Add filter pipeline
+   *
+   * POST |  foglamp/filter/{serviceName}/pipeline?allow_duplicates=true&append_filter=true
+   *
```
Comment (praveen-garg) [inline src/app/services/filter.service.ts]: remove this newline?

File: src/app/services/filter.service.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    return this.http.put(this.FILTER_URL + '/' + encodeURIComponent(serviceName)
+      + '/pipeline?allow_duplicates=true&append_filter=false', payload).pipe(
+        map(response => response),
+        catchError((error: Response) => observableThrowError(error)));
+  }
+
+  /**
+   * Delete filter
+   * DELETE | filter/{filterName}
+   *
```
Comment (praveen-garg) [inline src/app/services/filter.service.ts]: remove this extra line?

File: src/app/services/filter.service.ts [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    return this.http.get(this.CATEGORY_URL + '/' + encodeURIComponent(categoryName)).pipe(
+      map(response => response),
+      catchError((error: Response) => observableThrowError(error)));
+  }
+
+  /**
+   * Update filter pipeline
+   *
+   * PUT | filter/{serviceName}/pipeline?allow_duplicates=true&append_filter=false
+   *
```
Comment (praveen-garg) [inline src/app/services/filter.service.ts]: remove this extra line?

File: src/app/components/core/filter/add-filter-wizard/add-filter-wizard.component.ts [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+
+  /**
+   * Method to add filter
+   * @param payload  to pass in request
+   */
+  public addFilter(payload) {
+    this.filterService.saveFilter(payload)
+      .subscribe(
+        (data: any) => {
+          this.alertService.success(data.filter + ' filter added successfully.', true);
```
[the remark this answers is not in the document]
Comment (mshariq-nerd) [reply src/app/components/core/filter/add-filter-wizard/add-filter-wizard.component.ts]: ```{"filter": "S4", "description": "Configuration of 'S4' filter for plugin 'scale'", "value": {"plugin": {"default": "scale", "type": "string", "description": "Scale filter plugin", "value": "scale"}, "factor": {"default": "100.0", "type": "float", "description": "Scale factor for a reading value.", "value": "100.0"}, "offset": {"default": "0.0", "type": "float", "description": "A constant offset to add to every value.", "value": "0.0"}, "enable": {"default": "false", "type": "boolean", "description": "A switch that can be used to enable or disable execution of the scale filter.", "value": "false"}}}```

File: src/app/components/core/configuration-manager/view-config-item/view-config-item.component.ts (hunk 4 above)
Comment (mshariq-nerd) [reply src/app/components/core/configuration-manager/view-config-item/view-config-item.component.ts]: there is no key name `type ` in the object, so to put it there we need to define it.

Comment (praveen-garg) [reply src/app/components/core/configuration-manager/view-config-item/view-config-item.component.ts]: Okay, please add a comment `// there is no key 'type' in the object` before the line `d['type']`

File: src/app/components/core/filter/add-filter-wizard/add-filter-wizard.component.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (praveen-garg) [reply src/app/components/core/filter/add-filter-wizard/add-filter-wizard.component.ts]: Okay.

File: src/app/components/core/south/south-service-modal/south-service-modal.component.css [hunk 12]
```diff
@@ -18,3 +18,24 @@ label{
...
   display: none !important;
 }
+
+.dndPlaceholder {
+  min-height: 72px;
+  border: 1px dashed green;
+  background-color: rgba(0, 0, 0, 0.1);
+}
+
+.hide-text {
```
Comment (monika-Dianomic) [inline src/app/components/core/south/south-service-modal/south-service-modal.component.css]: Is this being used anywhere?

File: src/app/components/core/south/south-service-modal/south-service-modal.component.ts [hunk 13]
```diff
@@ -275,6 +345,21 @@ export class SouthServiceModalComponent implements OnInit, OnChanges {
     this.child.toggleModal(true);
   }
 
+
+  /**
+  * Open confirmation modal
+  * @param message   message to show on alert
```
Comment (monika-Dianomic) [inline src/app/components/core/south/south-service-modal/south-service-modal.component.ts]: there is no param in the method.

File: src/app/components/core/filter/add-filter-wizard/add-filter-wizard.component.ts [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+    if (event.target.value.trim().length > 0) {
+      this.isValidName = true;
+    }
+  }
+
+  public getInstalledFilterPlugins() {
+    this.filterService.getInstalledFilterPlugins().subscribe(
+      (data: any) => {
+        this.plugins = sortBy(data.plugins, p => {
+          return p.name;
```
Comment (praveen-garg) [inline src/app/components/core/filter/add-filter-wizard/add-filter-wizard.component.ts]: We should also handle the name starting with lowercase!

**Pull request merged**